crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,406 @@
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    pub fn reserve(&mut self, amount: usize) {
+        let available_space = self.capacity - self.len;
+        if available_space < amount {
+            self.grow(amount - available_space);
+        }
+    }
+
+    fn grow(&mut self, increment: usize) {
+        debug_assert!(self.item_layout.size() != 0);
+
+        let new_capacity = self.capacity + increment;
+        unsafe {
+            let new_data = if self.capacity == 0 {
+                std::alloc::alloc(Layout::from_size_align_unchecked(
+                    self.item_layout.size() * new_capacity,
+                    self.item_layout.align(),
+                ))
+            } else {
+                std::alloc::realloc(
+                    (*self.data.get()).as_ptr(),
+                    Layout::from_size_align_unchecked(
+                        self.item_layout.size() * self.capacity,
+                        self.item_layout.align(),
+                    ),
+                    self.item_layout.size() * new_capacity,
+                )
+            };
+
+            self.data = UnsafeCell::new(NonNull::new(new_data).unwrap());
+        }
+        self.capacity = new_capacity;
+    }
+
+    /// # Safety
+    /// `index` must be in bounds
+    /// Allows aliased mutable access to `index`'s data. Caller must ensure this does not happen
+    #[inline]
+    pub unsafe fn set_unchecked(&self, index: usize, value: *mut u8) {
+        let ptr = self.get_unchecked(index);
+        std::ptr::copy_nonoverlapping(value, ptr, self.item_layout.size());
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the value pointer points to data that matches self.item_layout
+    /// This [BlobVec] will take ownership of `value`'s data, which means the caller should ensure that data is not dropped early  
+    #[inline]
+    pub unsafe fn push(&mut self, value: *mut u8) {
+        let index = self.push_uninit();
+        self.set_unchecked(index, value);
+    }
+
+    /// increases the length by one (and grows the vec if needed) with uninitialized memory and returns the index
+    /// # Safety
+    /// the newly allocated space must be immediately populated with a valid value
+    #[inline]
+    pub unsafe fn push_uninit(&mut self) -> usize {
+        self.reserve(1);
+        let index = self.len;
+        self.len += 1;
+        index
+    }
+
+    /// # Safety
+    /// len must be <= capacity. if length is decreased, "out of bounds" items must be dropped. if length is increased,
+    /// newly added items must be immediately populated with valid values.
+    pub unsafe fn set_len(&mut self, len: usize) {
+        self.len = len;
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the type matches self.item_layout
+    #[inline]
+    pub unsafe fn push_type<T>(&mut self, mut value: T) {
```

Comment:
I would rather call it
```suggestion
    pub unsafe fn push_typed<T>(&mut self, mut value: T) {
```